src/solve3.F

```diff
@@ -1030,38 +1014,62 @@ subroutine solve3(nstep,rbufsz,num_soil_layers,                  &
             !$omp parallel do default(shared) private(i,j,k)
             !$acc parallel loop gang vector collapse(3) default(present) private(i,j,k)
             do k=1,nk
-            do j=1,nj
-            do i=1,ni
+            do j=1,nj+1
+            do i=1,ni+1
               th3d(i,j,k)=th3d(i,j,k)+dt*dpten(i,j,k,1)
               q3d(i,j,k,nqv)=q3d(i,j,k,nqv)+dt*dpten(i,j,k,2)
+              u3d(i,j,k)=u3d(i,j,k)+dt*dpten(i,j,k,3)
+              v3d(i,j,k)=v3d(i,j,k)+dt*dpten(i,j,k,4)
+              w3d(i,j,k)=w3d(i,j,k)+dt*dpten(i,j,k,5)
             enddo
             enddo
             enddo
             if(timestats.ge.1) time_parcels=time_parcels+mytime()
             call bcs_GPU(th3d)
             call bcs_GPU(q3d(ib,jb,kb,nqv))
+            call bcu_GPU(u3d)
+            call bcv_GPU(v3d)
+            call bcw_GPU(w3d,1)
 #ifdef MPI
             call sync()
             if(timestats.ge.1) time_lb=time_lb+mytime()
+            !----
             call comm_3s_start_GPU(th3d,rw31,rw32,re31,re32,   &
                                     rs31,rs32,rn31,rn32,reqs_y)
             call comm_3s_start_GPU(q3d(ib,jb,kb,nqv)  &
                        ,qw31(1,1,1,nqv),qw32(1,1,1,nqv),qe31(1,1,1,nqv),qe32(1,1,1,nqv)     &
                        ,qs31(1,1,1,nqv),qs32(1,1,1,nqv),qn31(1,1,1,nqv),qn32(1,1,1,nqv)     &
                        ,reqs_q(1,nqv) )
+            call comm_3u_start_GPU(u3d,uw31,uw32,ue31,ue32,   &
+                                       us31,us32,un31,un32,reqs_u)
+            call comm_3v_start_GPU(v3d,vw31,vw32,ve31,ve32,   &
+                                       vs31,vs32,vn31,vn32,reqs_v)
+            call comm_3w_start_GPU(w3d,ww31,ww32,we31,we32,   &
+                                       ws31,ws32,wn31,wn32,reqs_w)
+            !----
             call comm_3s_end_GPU(th3d,rw31,rw32,re31,re32,rs31,rs32,rn31,rn32,reqs_y)
             call comm_3s_end_GPU(q3d(ib,jb,kb,nqv)  &
                        ,qw31(1,1,1,nqv),qw32(1,1,1,nqv),qe31(1,1,1,nqv),qe32(1,1,1,nqv)     &
                        ,qs31(1,1,1,nqv),qs32(1,1,1,nqv),qn31(1,1,1,nqv),qn32(1,1,1,nqv)     &
                        ,reqs_q(1,nqv) )
+            call comm_3u_end_GPU(u3d,uw31,uw32,ue31,ue32,   &
+                                     us31,us32,un31,un32,reqs_u)
+            call comm_3v_end_GPU(v3d,vw31,vw32,ve31,ve32,   &
+                                     vs31,vs32,vn31,vn32,reqs_v)
+            call comm_3w_end_GPU(w3d,ww31,ww32,we31,we32,   &
+                                     ws31,ws32,wn31,wn32,reqs_w)
+            !----
 #endif
             !$omp parallel do default(shared) private(i,j,k)
             !$acc parallel loop gang vector collapse(3) default(present) private(i,j,k)
             do k=1,nk
-            do j=0,nj+1
-            do i=0,ni+1
+            do j=0,nj+2
+            do i=0,ni+2
```

Comment:
As @johnmauff pointed out, the change of loop bound here looked suspicious.

Reply:
We need @george-bryan to confirm here. While it seems safe (i.e., it's within the bounds of these arrays), I'm not certain why it was changed.